Add data manipulation documentation (#181)
cozy/cozy.github.io: src/tutorials/data/queries.md

However, if for some reasons you do not want to use cozy-client, it is possible to make queries on your own by directly requesting the [cozy-stack data API](https://docs.cozy.io/en/cozy-stack/data-system/).

In cozy-client, each query is defined as a [Query Definition](https://docs.cozy.io/en/cozy-client/api/cozy-client/#QueryDefinition). This API is helpful to easily chain Mango concepts (selectors, sorts, indexing, …), as we will see in the following sections.

In the following, we assume a DocType named `io.cozy.todos`  describing Todo lists.

Here is a basic example of a cozy-client query returning all the Todo lists in a Cozy:
```javascript
const queryDef = client.find("io.cozy.todos")
const docs = await client.queryAll(queryDef)
```

In this example, the `QueryDefinition` defines the query to get all the documents stored in the Todos database, expressed through `find("io.cozy.todos")`.

## Selectors

> To use the Mango query system, you define a `selector`, that allows to express criterias to filter the documents to return.

Reviewer: ```suggestion
To use the Mango query system, you define a `selector` expressing criterias to filter the documents to return.
```

### Selectors with cozy-client

> With cozy-client, a selector is expressed through the [`where`](https://docs.cozy.io/en/cozy-client/api/cozy-client/#querydefinitionwhereselector-querydefinition) method, that takes into input an object with the selector expression. For instance, this query means “the todos in 'sport’ category AND a title ‘Exercices’”: 

Reviewer: ```suggestion
With cozy-client, a selector is added to the query definition with the [`where`](https://docs.cozy.io/en/cozy-client/api/cozy-client/#querydefinitionwhereselector-querydefinition) method. It takes an object with the selector expression. For instance, this query means “the todos in 'sport’ category AND a title ‘Exercices’”: 
```

```javascript
const queryDef = client
  .find("io.cozy.todos")
  .where({
    "category": "sport",
    "title": "Exercices"
  })
```

`find` targets the Todos database, while `where` filters the documents to return.

> ℹ️ When running this query, an index will automatically be created on `category` and `title` fields. See the [index section](#index-fields) for more details about it. As mentioned in the [performances](#mango-performances), it is highly recommended to read it in order to avoid designing poorly efficient queries.

Reviewer: ```suggestion
ℹ️ When running this query, an index will automatically be created on the `category` and `title` fields. It is highly recommended to read the [index section](#index-fields) to avoid designing poorly efficient queries (see also [the section on mango performances](#mango-performances)). 
```

> ℹ️ The `$eq` and `$and` operators are implicit by default, which allow to write less verbose queries. Here is the explicit version of the previous query:

Reviewer: ```suggestion
ℹ️ The `$eq` and `$and` operators are implicit by default. It allows to write concise queries. Here is the explicit version of the previous query:
```

```javascript
const queryDef = client
  .find("io.cozy.todos")
  .where({
    "$and": [
      {
        "category": {
          "$eq": "sport"
      },
      {
        "title": {
          "$eq": "Exercices"
        }
      }
    ]
  })
```

### Mango operators

There are two kinds of mango operators:

- [Condition operators](https://docs.couchdb.org/en/2.3.1/api/database/find.html#condition-operators) (`$eq`, `$gt`, `$lt`, …)
- [Combinaison operators](https://docs.couchdb.org/en/2.3.1/api/database/find.html#combination-operators) (`$and` , `$or`, `$not`, …)

> For more details and examples of selectors, you can directly reach the [CouchDB documentation](https://docs.couchdb.org/en/2.3.1/api/database/find.html#selector-syntax).

Reviewer: ```suggestion
For more details and examples of selectors, you can directly read the [CouchDB documentation](https://docs.couchdb.org/en/2.3.1/api/database/find.html#selector-syntax).
```

### Mango performances

Depending on your queries and their complexity, performances can be dramatically impacted, by serveral orders of magnitude, especially for large databases, i.e. starting for thousands of documents.

> Thus, it is highly recommended to take some time to understand how to [index fields](#index-fields) and the related [performances](./advanced.md#indexes-performances-and-design).

Reviewer: ```suggestion
It is highly recommended to take some time to understand how to [index fields](#index-fields) and the related [performances](./advanced.md#indexes-performances-and-design).
```

"Thus" is a bit formal.

Author: I don't really see the problem with being formal :)

## Index fields

> In order to be able to efficiently filter documents through selectors and sort them, it is important to correctly index data. An index is a way to organize documents in order to retrieve them faster.

Reviewer: ```suggestion
In order to efficiently filter documents through selectors and sort them, it is important to correctly index data. An index is a way to organize documents in order to retrieve them faster.
```

> An important point to keep in mind is that **any field involved in a selector should be indexed** for performances. Otherwise, the documents filtering will be processed by CouchDB by fetching all the database documents and performing the selector in memory, which can be very time-consuming when the database grows. In some cases, it can even prevent the query to be run.

Author: Fair enough for the precision and the example, this rule should not be absolute

> ⚠️ For this reason, cozy-client automatically index fields declared in a `where` call. However, this currently only works for simple queries with implicit combinaison operators, i.e. where all the fields are declared at the first level of the object. For any query involving explicit operators, you can use the [`indexFields`](https://docs.cozy.io/en/cozy-client/api/cozy-client/#QueryDefinition+indexFields) method, see below for an example.

Reviewer: ```suggestion
⚠️ For this reason, cozy-client automatically indexes fields declared in a `where` call. However, this currently only works for simple queries with implicit combinaison operators, i.e. where all the fields are declared at the first level of the object. For any query involving explicit operators, you can use the [`indexFields`](https://docs.cozy.io/en/cozy-client/api/cozy-client/#QueryDefinition+indexFields) method, see below for an example.
```

Here, we explain how you can index fields to efficiently query documents. If you are interested on why indexing is important and how your performances can dramatically vary, you can go to the advanced explanations [here](./advanced.md#indexes-performances-and-design).

With mango queries, it is required to index documents fields that are necessary to run the queries, either because they are involved in the selector or the sort.
  
### Index with cozy-client

> With cozy-client, the developer does not have to deal with the index creation: it is expressed through `indexFields`:

Reviewer: NIt: Here you switch between "the developer" and "you".

```javascript
const queryDef = client
  .find("io.cozy.todos")
  .where({
    "category": "sport"
  })
.indexFields(["category"])
```

> Thanks to this, cozy-client will automatically create an index on `category` before running the query for the first time and explicity tell CouchDB to use it.

Reviewer: ```suggestion
Thanks to this, cozy-client automatically creates an index on `category` before running the query for the first time and explicity tell CouchDB to use it.
```

💡 In this example, the `indexField` method is not mandatory: cozy-client is smart enough to automatically index `category`  as it is involved in the selector. However, this automatic indexing only works for simple queries, thus, it is encouraged to explicity declare which fields to index.

> ⚠️ CouchDB evaluates if a new document should be indexed by checking if its fields match an existing index. It means that **all the indexed fields must exist in the document** to index it. It can be problematic for queries that needs to check if a particular field exists or not. We detail this behaviour and workarounds in [this section](./advanced.md#indexes-why-is-my-document-not-being-retrieved).

Reviewer: ```suggestion
⚠️ CouchDB evaluates if a new document should be indexed by checking if its fields match an existing index. It means that **all the indexed fields must exist in the document** to index it. It can be problematic for queries that needs to check if a particular field exists or not. We detail this behaviour and workarounds in the ["why is my document not retrieved" section](./advanced.md#indexes-why-is-my-document-not-being-retrieved).
```

ℹ️ If no index is used to run the query, the query response will include a `warning: no matching index found, create an index to optimize query time`. It means that your index is not properly defined, probably because some fields involved in the `where`  selector are not indexed. Note that if you use a `sortBy`, you also need to index the fields involved, as explained in [this section](#sort-data).

### Index update

CouchDB updates the Mango indexes when the data is read, but not on writes. This implies that if one creates an index and inserts documents, the index will never be updated until a `find` query on these documents is performed. 

> ⚠️ This implies that a query performed after many writes can take a certain time to complete, as it will update the index before returning any result. Likewise, the first query after an index creation will include the index build. Hence, developers should keep this behaviour in mind when designing applications and typically might need to handle query latency. 

Reviewer: What does it mean to "handle query latency" ? Show a spinner ? Have an intermediate cache ?

Author: I don't know, for me it is up to the app developer

> ℹ️ Starting from CouchDB 3.0.0, a [background indexing](https://docs.couchdb.org/en/3.0.0/config/indexbuilds.html) is implemented that aims to avoid this latency by forcing the index to update in background after writes operations.

Reviewer: ```suggestion
ℹ️ Starting from CouchDB 3.0.0, [background indexing](https://docs.couchdb.org/en/3.0.0/config/indexbuilds.html) is implemented. It aims to avoid this latency by forcing the index to update in background after writes.
```

### Index with several fields

> Here is how to declare a query asking for todos having a “work” category OR a creation date starting from 2019:

Reviewer 2: I think you want `AND`, not `OR` here. With `OR`, the index is useless: you can have a document with a creation date in 2019 with no category that should be in the results but it cannot be indexed due to the absence of category.

Author: I assumed all the fields are mandatory and filled. Actually I wanted a `or` operator to have at least one example with an explicit combinaison operator

Reviewer 2: But CouchDB won't check if the fields are mandatory and filled, and thus won't use the index at all.

```javascript
const queryDef = client
  .find("io.cozy.todos")
  .where({
    "$or": [
      {
        "created_at": {
          "$gt": "2019-01-01"
      },
      {
        "category": "work"
      }
    ]
  })
  .indexFields(["category", "created_at"])
```

> ⚠️ The order in the `indexFields` array can be very important for performances, especially when there are both range and equality operators: you must always index the field involved in the equality operator first. To understand this, one needs to understand some advanced concepts about CouchDB indexes, explained in [this section](#indexes-performances-and-design).

Reviewer: ```suggestion
⚠️ The order in the `indexFields` array can be very important for performances, especially when there are both range and equality operators: you must always index the field involved in the equality operator first. To understand this, it helps to understand concepts about CouchDB indexes, explained in [this section](#indexes-performances-and-design).
```

## Sort data with Mango

It is possible to express the order of the documents returned by a query, by using the `sort`  array. 

> The sort is always done on a document field, by ascending (`asc`) or descending (`desc`) order, ascending being the default.  One can specify several fields to sort,  by following this structure for each field:

Reviewer: ```suggestion
The sort is always done on a document field, by ascending (`asc`) or descending (`desc`) order, ascending being the default.  It's possible to specify several fields to sort, by following this structure for each field:
```

`{"fieldName": "asc"|"desc"}`

For instance:

`[{"category": "asc"}, {"created_at": "asc"}]`

Here, the returned documents will be first sorted by category, and then by creation date.

⚠️ It is not possible to mix the sort order on different fields, i.e. all the fields involved in the sort must be either `asc`, either `desc` but not a mix of both.

See [this section](#sort-in-couchdb) for more info about the sort order on data types.

### Sort with cozy-client

> With cozy-client, a sort is expressed through the `sortBy` method, that takes into input an array containing the fields to sort and the direction. For instance:

Reviewer: ```suggestion
With cozy-client, a sort is expressed through the `sortBy` method. It takes as argument an array containing the fields to sort and the direction. For instance:
```

```javascript
const queryDef = client
  .find("io.cozy.todos")
  .where({
    "category": "sport",
  })
  .indexFields(["category", "created_at"])
  .sortBy[{"category": "asc"}, {"created_at": "asc"}]
```

> ⚠️ Any field involved in the `sortBy` must be indexed: otherwise, CouchDB would be forced to scan all the documents in the database, which is not a good for scalability.

Reviewer 2: No, CouchDB won't be forced to scan all the documents, only to sort the results in memory, which is also bad, but not on the same way: if you are sure that to have few results, you can do that (even if the database has millions of records).

Author: Yes, I meant all the documents involved in the query, I'll add the precision

> ⚠️ Inversely, when using a sort, any indexed field must appear either in the `where` selector or the `sortBy`.  

Reviewer 2: Not `or`. It is not OK to index on `category, created_at`, have a selector on `category` and sort on `created_at` (CouchDB is a bit stupid).

Author: I just checked, and actually, when using a `category, created_at` index, it is necessary to either have a selector on both fields, either have a selector on `category` and a sort on `created_at`. Otherwise, the index is not used at all.

> ⚠️ At least one of the sort fields must be included in the `where` selector.

Reviewer 2: No, if it is the second field of `sort` that is in the `where`, it is useless. We should say that « At least the first field of the sort must be included in the `where` selector. But even that is not totally true: if you want all the records from a database sorted by a given criterion, you should index the database with this criterion and makes paginated requests with no `where` selector.

Author: > At least one of the sort fields is included in the selector.

This comes from the [CouchDB doc](https://docs.couchdb.org/en/stable/api/database/find.html#sort-syntax). I agree that you need the first field to be indexed, but I'm sure I understand "makes paginated requests with no where selector": why not use a selector? You might want to filter on a condition and a sort on it

Reviewer 2: I was thinking of a case like "I need all the bank accounts sorted by name". We don't want to filter on a condition. But the CouchDB doc is right: if you have an index on name and you make a request with no selector and a sort on name, the index won't be used. A selector that says `{ "name": "$exists" }` is required to make it possible to use the index.

See also the examples provided in the [PouchDB documentation](https://pouchdb.com/guides/mango-queries.html#more-than-one-field).

## Pagination

When dealing with queries returning a lot of documents, e.g. thousands, it might be necessary to paginate the results to avoid huge network payloads and having to load everything in memory, both on the server and client sides.

The pagination consists of splitting the results in “pages”: if a query matches 100.000 documents, one can paginate it, for example by actually running 1000 queries returning 100 documents each. 
By doing so, the server only keeps 100 documents in memory for each query, and the client can control the data flow, for example, by implementing a “Load more” button that actually runs a paginated query and loads the 100 next documents.

> With mango queries, the recommanded method to paginate is through the `bookmark` parameter, that works exactly as it name suggests: it consists of a string returned by CouchDB that corresponds to the page position in the index. Just like an actual bookmark is used to mark a particular page in a book.

Reviewer: ```suggestion
With mango queries, the recommended method to paginate is through the `bookmark` parameter, that works exactly as it name suggests: it consists of a string returned by CouchDB that corresponds to the page position in the index. Just like an actual bookmark is used to mark a particular page in a book.
```

```javascript
const queryDef = client.find("io.cozy.todos")
const docs = await client.queryAll(queryDef)
```

Note that it will actually automatically paginate if the total number of documents to return is superior to the `limit` (100 by default): this avoids loading too many documents in memory on the server side.
However, it is possible to overcome this pagination by passing `limit: null`: in this case, cozy-client will request the [`_all_docs` endpoint](https://docs.couchdb.org/en/stable/api/database/bulk-api.html#db-all-docs) and retrieve all the documents in one query. This endpoint is automatically created by CouchDB and is actually a [view](#views) on the `_id` key.

Thus, doing this will actually query the `_all_docs` endpoint and return all the documents in one request:

```javascript
const queryDef = client.find("io.cozy.todos")
const docs = await client.query(queryDef, {limit: null})
```

⚠️ When querying this endpoint, the response includes the [design docs](https://docs.couchdb.org/en/master/ddocs/index.html), which are the Mango indexes and views definitions.  Those documents are automatically filtered for the paginated queries.

> ⚠️ This method is faster than the pagination as it avoids to make several server requests. However, if there are many documents to return, `_all_docs` queries can take a lot of time to complete and consume server resources. Hence, you should be cautious when using this route.

Reviewer 2: Maybe we should add that the requests can even timeout.

## Sort in CouchDB

When dealing with various types of data, the ascending sort order is the following:

- null
- booleans
- numbers
- strings: “a” < “A” < ”aa” < “b” …
- arrays
- objects

See the [CouchDB documentation](https://docs.couchdb.org/en/2.3.1/ddocs/views/collation.html#collation-specification) for more details on the sort orders.

ℹ️ The sort order is the same for Mango queries and views. 

### Comparison of strings

Comparison of strings is done using ICU which implements the Unicode Collation Algorithm, giving a dictionary sorting of keys. This can give surprising results if you were expecting ASCII ordering. Note that:

- All symbols sort before numbers and letters (even the “high” symbols like tilde, `0x7e`)
- Differing sequences of letters are compared without regard to case, so `a < aa` but also `A < aa` and `a < AA`
- Identical sequences of letters are compared with regard to case, with lowercase before uppercase, so `a < A`.

## Views

> ⚠️ In Cozy, we disabled by default the possibility to create views from applications. Consequently, [cozy-client](https://github.com/cozy/cozy-client) does not support views creation, because they are disabled for applications. See the [view performances](./advanced.md#views-performances) section to find out why. 

Reviewer: ```suggestion
⚠️ In Cozy, we disabled by default the possibility to create views from applications. Consequently, [cozy-client](https://github.com/cozy/cozy-client) does not support view creation, because they are disabled for applications. See the [view performances](./advanced.md#views-performances) section to find out why. 
```

[CouchDB views](https://docs.couchdb.org/en/2.3.1/ddocs/views/index.html) are another way to express queries in Cozy, used by the [cozy-stack](https://github.com/cozy/cozy-stack/blob/331f6f39402dc7ec38ecf2f5ff408a1b5cf3d730/pkg/couchdb/index.go#L48-L229). It consists of creating [map-reduce](https://en.wikipedia.org/wiki/MapReduce) functions, where the `map` is used to qualify which documents should be indexed, and the optional `reduce`  can be used to compute aggregate functions on them.

Views are much more flexible than Mango queries, as the `map`  is a user-defined function (UDF) that can be written in JavaScript.

For example:

```javascript
function(doc) {
  if(doc.category && doc.created_at) {
    emit(doc.category, doc.created_at);
  }
}
```

A `map` function must always produce an `emit` , with the key and the value respectively in first and second argument.

> ℹ️ Behind the hoods, views are represented exactly like the Mango queries: both uses [B+ Tree](./advanced.md#indexes-concepts), with values stored in the leafs.

Reviewer: ```suggestion
ℹ️ Behind the hoods, views are represented exactly like the Mango queries: both use [B+ Tree](./advanced.md#indexes-concepts), with values stored in the leafs.
```

### Query a view

> Just like the Mango queries, views are particularly suited for equality and range queries. See the [CouchDB documentation](https://docs.couchdb.org/en/2.3.1/api/ddoc/views.html) to know more about the API.

Reviewer: ```suggestion
Just like Mango queries, views are particularly suited for equality and range queries. See the [CouchDB documentation](https://docs.couchdb.org/en/2.3.1/api/ddoc/views.html) to know more about the API.
```

To find a specific document, use the  `key` parameter:

`key="sport"`

To find a range, combine the `startkey` and `endkey` parameters:

> `startkey="category1"&endkey="category9"`

Reviewer 2: I'm not sure it is a very good example, as `category42` is between `category1` and `category9`. Maybe add padding: `category0001` and `category0009`.
